### ISSUE TITLE
[FEATURE] Support conversional search in ML Inference Search Response Processor with memory 

### ISSUE THREAD

**Issue by mingshl** [opensearch-project/ml-commons#3242]
**Is your feature request related to a problem?**
to support conversational search, when sending the request to the remote model, we not only needs to send the questions, but also the historical context.

For example, 
OpenAI API: 
```
messages = [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "Hello, how are you?"},
    {"role": "assistant", "content": "I'm doing well, thank you. How can I assist you today?"},
    {"role": "user", "content": "What's the weather like?"}
]
```

Bedrock Claude API: 
```
conversation = """Human: Hello, I'd like to know about the weather in New York.
Assistant: Hello! I'd be happy to help you with information about the weather in New York. However, I don't have access to real-time weather data. Could you please specify what kind of weather information you're looking for? For example, are you interested in the current weather, a forecast for the coming days, or general climate information about New York?
```

in ml inference search response processor, introduce a new parameter, "conversational_search", to be true or false. when it's true, and the input_map config to read the memory id from query extension. 

 ```
{
  "ml_inference": {
    "model_id": "<model_id>", 
    "conversational_search": true
    "function_name": "<function_name>",
    "full_response_path": "<full_response_path>",
    "model_config":{
      "<model_config_field>": "<config_value>"
    },
    "model_input": "<model_input>",
    "input_map": [
      {
        "memory_id": "$._query.ext.ml_inference.memory_id"
        "content": "$._query.ext.ml_inference.question"
      }
    ],
    "output_map": [
      {
        "<new_document_field>": "<model_output_field>"
      }
    ],
    "override": "<override>",
    "one_to_one": false
  }
}
 ```

when search in query, users can use the ml inference search extension to ask question. 

```
GET /my_rag_test_data/_search?search_pipeline=rag_pipeline
{
  "query": {
    "match": {
      "text": "Abraham Lincoln"
    }
  },
  "ext": {
    "ml_inference": {
      "llm_question": "Was Abraham Lincoln a good politician",
      "memory_id": "iXC4bI0BfUsSoeNTjS30"
    }
  }
}
```

**What alternatives have you considered?**

**Do you have any additional context?**

**Comment by Distorted-Dundar** [comment]
Hey,

Is the memory id an index? If so what are some responsibilities of this index it looks like the ml inference processor will load the index and store info in the index on return is there a restriction on this memory id or can any other index be used?

How do we know we should clean up these memories after sometime?

**Comment by austintlee** [comment]
I worry that this might blur the line between the ML response processor and the existing RAG processor.  We may be adding too much to the ML inference processor interface.

**Comment by mingshl** [comment]
> I worry that this might blur the line between the ML response processor and the existing RAG processor. We may be adding too much to the ML inference processor interface.

Hi @austintlee , we are working on OpenSearch Flow Project, you can refer here https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/tutorial.md for the tutorial.  

This OpenSearch Flow is aiming at using ML Inference Processors (ingest/search) as a generic processor to run inference during ingest and search in a workflow to simplify set up and configurations. Of course, if users are familiar with Rag processors or others existing processors, users can use other processors as well, there are drop down options in the processors option that user can pick and bindle, it's up to the users choice for their use cases.

**Comment by mingshl** [comment]
> Hey,
> 
> Is the memory id an index? If so what are some responsibilities of this index it looks like the ml inference processor will load the index and store info in the index on return is there a restriction on this memory id or can any other index be used?
> 
> How do we know we should clean up these memories after sometime?

yes the memory would be story in index. Indeed the memory and message API were releases. checkout this doc https://opensearch.org/docs/latest/ml-commons-plugin/api/memory-apis/get-memory/ https://opensearch.org/docs/latest/ml-commons-plugin/api/memory-apis/get-message/
